packages/server/src/ws-connection.ts

```diff
@@ -0,0 +1,184 @@
+  isServerEventDataState,
+  lookupValue,
+  response,
+} from '@rise-tools/react'
+import { WebSocket } from 'ws'
+import { z } from 'zod'
+
+import { findModel, getModelState } from './model-utils'
+import { AnyModels, ValueModel } from './types'
+
+const serverSubscribeMessageSchema = z.object({
+  $: z.literal('sub'),
+  keys: z.array(z.string()),
+})
+const serverUnsubscribeMessageSchema = z.object({
+  $: z.literal('unsub'),
+  keys: z.array(z.string()),
+})
+const serverEventMessageSchema = z.object({
+  $: z.literal('evt'),
+  event: z.object({
+    target: z.object({
+      key: z.string().optional(),
+      component: z.string(),
+      propKey: z.string(),
+      path: z.array(z.string().or(z.number())),
+    }),
+    dataState: z.any(),
+    payload: z.any(),
+  }),
+  key: z.string(),
+})
+
+const serverMessageSchema = z.discriminatedUnion('$', [
+  serverSubscribeMessageSchema,
+  serverUnsubscribeMessageSchema,
+  serverEventMessageSchema,
+])
+
+export type WSServerContext = {
+  models: AnyModels
+  getModel: (key: string) => ValueModel<unknown> | (() => unknown) | undefined
+  clientIdIndex: number
+  clientSenders: Map<string, (value: any) => void>
+  clientSubscribers: Map<string, Map<string, () => void>>
+}
+
+export function createWSServerContext(models: AnyModels): WSServerContext {
+  function getModel(key: string) {
+    const path = key.split('/')
+    const model = findModel(models, path)
+    return model
+  }
+  return {
+    models,
+    getModel,
+    clientIdIndex: 0,
+    clientSenders: new Map<string, (value: any) => void>(),
+    clientSubscribers: new Map<string, Map<string, () => void>>(),
+  }
+}
+
+export function connectWebSocket(context: WSServerContext, ws: WebSocket) {
+  const clientId = `c${context.clientIdIndex}`
+  context.clientIdIndex += 1
+  // console.log(`Client ${clientId} connected`)
+
+  const { clientSenders, clientSubscribers, models, getModel } = context
+
+  clientSenders.set(clientId, function sendClient(value: any) {
+    ws.send(JSON.stringify(value))
+  })
+
+  ws.addEventListener('close', function close() {
+    clientSenders.delete(clientId)
+    // console.log(`Client ${clientId} disconnected`)
+  })
+
+  function handleSubKey(key: string) {
+    // console.log(`${clientId} sub to key ${key}`)
+    async function send() {
+      const sender = clientSenders.get(clientId)
+      if (!sender) {
+        return
+      }
+      const model = getModel(key) as ValueModel<unknown> | undefined
+      const val = typeof model === 'function' ? model() : model?.get()
+      // console.log('sending', model, key, val)
+      sender({
+        $: 'up',
+        key,
+        val,
+      })
+    }
+    const handlers =
+      clientSubscribers.get(key) ||
+      (() => {
+        const handlers = new Map<string, () => void>()
+        clientSubscribers.set(key, handlers)
+        const model = getModel(key)
+        if (model && typeof model === 'object') {
+          model.subscribe(() => {
```

Comment:
Should we clean-up this subscription as well?

Reply:
true true! will try to fix this